### PR TITLE
chore: firefly runtime

### PR DESCRIPTION
**Pull request by guanbinrui**
## Description

Closes MF-0000

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/web3-providers/src/Web3/EVM/providers/CustomEvent.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import { first } from 'lodash-es'
+
+export class EVMCustomEventProvider extends BaseEVMWalletProvider {
+    constructor() {
+        super(ProviderType.CustomEvent)
+    }
+
+    override async setup() {
+        // @ts-expect-error TODO: define the custom event
+        document.addEventListener(
```
Comment (Jack-Works) [inline packages/web3-providers/src/Web3/EVM/providers/CustomEvent.ts:17]: please only enable this one on the domain you needed

File: packages/shared-base/src/SingletonModal/index.ts [hunk 2]
```diff
@@ -116,6 +116,16 @@ export class SingletonModal<
             this.onAbort = (error) => reject(error)
         })
     }
+
+    __unsafe_overwrite_methods__(methods: {
```
Comment (Jack-Works) [inline packages/shared-base/src/SingletonModal/index.ts:120]: can you document the purpose of this function?

Comment (Jack-Works) [reply packages/shared-base/src/SingletonModal/index.ts:120]: hello? cc @UncleBill

Comment (UncleBill) [reply packages/shared-base/src/SingletonModal/index.ts:120]: It might be written by @guanbinrui

**Pull request merged**